<?xml version="1.0"?>
<launch>

  <!-- Parameters -->
  <arg name="model" default="burger"/>
  <arg name="use_sim_time" default="true" />
  <arg name="shelfino_name" default="limo1" />
  <arg name="use_gui" default="true" />
  <arg name="use_rviz" default="true" />
  <!--arg name="rviz_config_file" default="$(find loco_planning)/rviz/model_display.rviz" /-->
  <!--arg name="gazebo_world_file" default="$(find loco_planning)/worlds/empty.world" /-->
  <arg name="spawn_shelfino" default="false" />
  <arg name="shelfino_init_x" default="0.0" />
  <arg name="shelfino_init_y" default="0.0" />
  <arg name="shelfino_init_yaw" default="0.0" />

  <!--arg name="nav_param_file" default="$(find shelfino_navigation)/config/shelfino.yaml" /-->
  <arg name="nav_rviz_config_file" default="$(find map_pkg)/rviz/shelfino_nav.rviz" />
  <arg name="map_env_params_file" default="$(find map_pkg)/config/map_config.yaml" />
  <arg name="gen_map_params_file" default="$(find map_pkg)/config/full_config.yaml" />
  <arg name="elements_models_path" default="$(find map_pkg)/models" />
  <arg name="gazebo_models_path" default="$(find map_pkg)/worlds" />
  
  
   <arg name="victims_activated" default="true" />
  <arg name="generate_new_map_config" default="true" />

  <arg name="x_pos" default="-2.0"/>
  <arg name="y_pos" default="-0.5"/>
  <arg name="yaw_pos" default="0.0"/>


  <!-- Set parameters -->
  <param name="use_sim_time" value="$(arg use_sim_time)" />



  <!-- Generate the config file  and load params into param server-->
  <node pkg="map_pkg" type="generate_config_file.py" name="generate_config_file" output="screen">
    <param name="map_env_params_file" value="$(arg map_env_params_file)" />
    <param name="gen_map_params_file" value="$(arg gen_map_params_file)" />
    <param name="generate_new_config" value="$(arg generate_new_map_config)" />
    <param name="victims_activated" value="$(arg victims_activated)" />
  </node>



  <!-- Spawn map publishes borders and obstacles-->
  <include file="$(find map_pkg)/launch/spawn_map.launch">
  <param name="map_env_params_file" value="$(arg map_env_params_file)" />
    <param name="generate_new_config" value="$(arg generate_new_map_config)" />
    <arg name="gen_map_params_file" value="$(arg gen_map_params_file)" />
    <arg name="gazebo_models_path" value="$(arg gazebo_models_path)" />
    <arg name="elements_models_path" value="$(arg elements_models_path)" />
    <arg name="victims_activated" value="$(arg victims_activated)" />
  </include>


<!-- Create PGM map based on published borders and obstacles NNEDE only for nav2 it generates the dynamic_map.yaml and runs the mapserver-->
  <node pkg="map_pkg" type="create_map_pgm.py" name="create_map_pgm" output="screen">
    <rosparam command="load" file="$(arg map_env_params_file)" />
  </node>



  <!--Spawn turtle bot robot and description-->
  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <!--arg name="world_name" value="$(find turtlebot3_gazebo)/worlds/turtlebot3_world.world"/-->
    <arg name="paused" value="false"/>
    <arg name="use_sim_time" value="true"/>
    <arg name="gui" value="true"/>
    <arg name="headless" value="false"/>
    <arg name="debug" value="false"/>
  </include>



  <!--spawn the limo robot (default limo1)-->
  <include file="$(find limo_description)/launch/upload.launch"/> 
  <node name="spawn_model" pkg="gazebo_ros" type="spawn_model" args="-urdf -param robot_description -model limo1 -x $(arg x_pos) -y $(arg y_pos)  -Y $(arg yaw_pos)"/>



  <!-- AMCL Localization -->
  <!--node pkg="amcl" type="amcl" name="amcl" output="screen"/-->

 <!-- this publishes in ground_truth topic the TF  from odom to base_link needed if you do not use controller-->
 <node pkg="map_pkg" type="ground_truth_odom.py" name="ground_truth_odom" output="screen">
    <param name="robot_name" value="$(arg shelfino_name)" />
      <param name="rate" value="100." />
 </node>
 <!-- this emulates perfect no drift condition matching odom (that usually drifts) with map -->
  <node pkg="tf" type="static_transform_publisher" name="map_to_odom_static_broadcaster"
      args="0 0 0 0 0 0 map $(arg shelfino_name)/odom 100" />


  <!-- Publish robot joint states -->
<node pkg="joint_state_publisher" type="joint_state_publisher" name="joint_state_publisher">
  <param name="rate" value="100.0" />
</node>
 <node pkg="robot_state_publisher" type="robot_state_publisher" name="robot_state_publisher">
    <param name="publish_frequency" type="double" value="100.0" />
 </node>

 
  <!-- RViz (optional) -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find map_pkg)/rviz/conf.rviz" />



</launch>
